source/custom/nixnetsocket_service.custom.cpp

```diff
@@ -1,2 +1,94 @@
+#include <nixnetsocket/nixnetsocket_service.h>
+
+#include <atomic>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include "xnetsocket_converters.h"
+
 namespace nixnetsocket_grpc {
+
+// Returns true if it's safe to use outputs of a method with the given status.
+inline bool status_ok(int32 status)
+{
+  return status >= 0;
+}
+
+//---------------------------------------------------------------------
+//---------------------------------------------------------------------
+::grpc::Status NiXnetSocketService::GetSocketOption(::grpc::ServerContext* context, const GetSocketOptionRequest* request, GetSocketOptionResponse* response)
+{
+  if (context->IsCancelled()) {
+    return ::grpc::Status::CANCELLED;
+  }
+  try {
+    auto socket_grpc_session = request->socket();
+    nxSOCKET socket = session_repository_->access_session(socket_grpc_session.id(), socket_grpc_session.name());
+    int32_t level = request->level();
+    int32_t opt_name;
+    switch (request->opt_name_enum_case()) {
+      case nixnetsocket_grpc::GetSocketOptionRequest::OptNameEnumCase::kOptName: {
+        opt_name = static_cast<int32_t>(request->opt_name());
+        break;
+      }
+      case nixnetsocket_grpc::GetSocketOptionRequest::OptNameEnumCase::kOptNameRaw: {
+        opt_name = static_cast<int32_t>(request->opt_name_raw());
+        break;
+      }
+      case nixnetsocket_grpc::GetSocketOptionRequest::OptNameEnumCase::OPT_NAME_ENUM_NOT_SET: {
+        return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for opt_name was not specified or out of range");
+        break;
+      }
+    }
+
+    response->mutable_optval()->set_opt(request->opt_name());
+    nxsocklen_t optlen{};
+    int32_t status;
+    switch (request->opt_name()) {
```

Comment:
This is ugly, but seemed like the most straightforward way to allocate the void* passed to the library method and set it back the output response. I tried for a bit with `convert_to_grpc` but wasn't sure how to pass the reference to the data to be filled by GetSocketOption in the first place without doing something like what's seen here.